lib/ash/type/union.ex

```diff
@@ -695,11 +695,13 @@ defmodule Ash.Type.Union do
         new_values,
         constraints
       ) do
-    if Enum.any?(constraints[:types] || [], fn {_name, config} ->
-         Ash.Type.prepare_change_array?(config[:type])
-       end) do
+    if !is_nil(new_values) &&
```

Comment:
Hm...actually I approved this, but I'm thinking that we shouldn't call `prepare_change_array` with a `nil` value ever, and that is probably what changed here is that somehow we started doing this. I think we should solve it at the call site instead of here perhaps.

Reply:
@zachdaniel ok, so you're thinking we'd check here?

https://github.com/ash-project/ash/blob/49958975a18758bdc4ee74e8bdb67028cff58e68/lib/ash/type/type.ex#L593

and here for handle_change_array?

https://github.com/ash-project/ash/blob/49958975a18758bdc4ee74e8bdb67028cff58e68/lib/ash/type/type.ex#L572

Reply:
Actually...I think I might be changing my mind here. It makes more sense to let the underlying types handle `nil`. The problem with this is that if the value is `nil`, we still need to attempt to call the underlying `prepare_change_array` with the `nil` value. So instead of skipping, we need to skip the reindexing logic and cast each group of old values against a `nil` value for the type.

It's more complex, but also more accurate.

Reply:
@zachdaniel I've removed this from this PR, since as I mentioned it wasn't required for our current issue. I'll put that in a different PR.